File: SuperBackup/SuperBackup/SettingParser.cpp
#include "SettingParser.h"
#include "Setting.h"
#include <fstream>
#include <sstream>
#include <stdexcept>

SettingParser::SettingParser(const std::string filename)
{
	this->filename = filename;
	parseContent();
	contentToSettings();
}

SettingParser::~SettingParser()
{
	// @TODO: Clear map & vectors.
}

bool SettingParser::keyExists(const std::string key) const
{
	return contents.find(key) != contents.end();
}

std::vector<std::string> SettingParser::getKey(const std::string key) const
{
	if (!keyExists(key))
	{
		throw std::invalid_argument("Key does not exist.");
	}

	return contents.at(key);
}

void SettingParser::parseContent()
{
	std::ifstream infile(this->filename);
	std::string line;

	while (std::getline(infile, line))
	{
		if (validLine(line))
		{
			std::istringstream is_line(line);
			std::string key;

			if (std::getline(is_line, key, '=')) // Extract and store key from line.
			{
				std::string value;
				
				if (key[0] == '#' || key[0] == ';') // Ignore comments
				{
					continue;
				}

				if (std::getline(is_line, value))
				{
					// Store key & value
					contents[key].push_back(value);
				}
			}
		}
	}

	infile.close();
}

void SettingParser::contentToSettings()
{
	for (auto const &x : contents)
	{
		for (auto const &vec : contents[x.first])
		{
			Setting s;

			// Check if filetype or path and create Setting.
			if (x.first.find("dir") != std::string::npos || x.first.find("directory") != std::string::npos)
			{
				s.Initialize(DIRECTORY, vec.c_str());
				settings.AddSetting(s);
			}
			else if (x.first.find("filetype") != std::string::npos)
			{
				s.Initialize(FILETYPE, vec.c_str());
				settings.AddSetting(s);
			}
		}
	}
}

bool SettingParser::validLine(const std::string &line) const
{
	if (line[0] != '#' || line[0] != ';') { // If not comments.
		if (line.find("=") != std::string::npos) // If contains an equal sign.
		{
			return true;
		}
	}

	return false;
}

File: src/me/oskarmendel/superbackup/setting/SettingParser.java
package me.oskarmendel.superbackup.setting;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * 
 * @author Oskar
 *
 */
public class SettingParser {
	
	private String filename;
	private Map<String, ArrayList<String>> contents;
	private SettingContainer settings;
	
	public SettingParser(String filename) {
		this.filename = filename;
		this.contents = new HashMap<String, ArrayList<String>>();
		parseContent();
		contentToSettings();
	}
	
	public boolean keyExists(String key) {
		return contents.containsKey(key);
	}
	
	ArrayList<String> getKey(String key) {
		if (!contents.containsKey(key)) {
			throw new IllegalArgumentException("Specified key does not exist.");
		}
		
		return contents.get(key);
	}
	
	SettingContainer getSettings() {
		return settings;
	}
	
	private void parseContent() {
		try {
			BufferedReader br = new BufferedReader(new InputStreamReader(new FileInputStream(filename)));
			
			String line;
			while ((line = br.readLine()) != null) {
				if (validLine(line)) {
					String key;
					String value;
					int delim = line.indexOf('=');
					
					key = line.substring(0, delim).trim();
					value = line.substring(delim+1).trim();
					
					if (contents.get(key) == null) {
						contents.put(key, new ArrayList<String>());
					}
					
					contents.get(key).add(value);
				}
			}
			
			br.close();
			
		} catch(IOException e) {
			e.printStackTrace();
		}
	}
	
	private void contentToSettings() {
		for (String key : contents.keySet()) {
			for (String value : contents.get(key)) {
				
				if (key.equals("directory") || key.equals("dir")) {
					Setting s = new Setting(SettingType.DIRECTORY, value);
					settings.addSetting(s);
				} else if (key.equals("filetype")) {
					Setting s = new Setting(SettingType.FILETPYE, value);
					settings.addSetting(s);
				}
				
			}
		}
	}
	
	private boolean validLine(String line) {
		if (line.charAt(0) != '#' || line.charAt(0) != ';') {
			if (line.contains("=")) {
				return true;	
			}
		}
		
		return false;
	}
}
